crates/ra_hir/src/ty/method_resolution.rs

```diff
@@ -176,27 +188,51 @@ pub(crate) fn iterate_method_candidates<T>(
     db: &impl HirDatabase,
     resolver: &Resolver,
     name: Option<&Name>,
-    mut callback: impl FnMut(&Ty, Function) -> Option<T>,
+    mode: LookupMode,
+    mut callback: impl FnMut(&Ty, AssocItem) -> Option<T>,
 ) -> Option<T> {
-    // For method calls, rust first does any number of autoderef, and then one
-    // autoref (i.e. when the method takes &self or &mut self). We just ignore
-    // the autoref currently -- when we find a method matching the given name,
-    // we assume it fits.
+    let krate = resolver.krate()?;
+    match mode {
+        LookupMode::MethodCall => {
+            // For method calls, rust first does any number of autoderef, and then one
+            // autoref (i.e. when the method takes &self or &mut self). We just ignore
+            // the autoref currently -- when we find a method matching the given name,
+            // we assume it fits.
 
-    // Also note that when we've got a receiver like &S, even if the method we
-    // find in the end takes &self, we still do the autoderef step (just as
-    // rustc does an autoderef and then autoref again).
+            // Also note that when we've got a receiver like &S, even if the method we
+            // find in the end takes &self, we still do the autoderef step (just as
+            // rustc does an autoderef and then autoref again).
 
-    let krate = resolver.krate()?;
-    for derefed_ty in autoderef::autoderef(db, resolver, ty.clone()) {
-        if let Some(result) = iterate_inherent_methods(&derefed_ty, db, name, krate, &mut callback)
-        {
-            return Some(result);
+            for derefed_ty in autoderef::autoderef(db, resolver, ty.clone()) {
+                if let Some(result) =
+                    iterate_inherent_methods(&derefed_ty, db, name, mode, krate, &mut callback)
+                {
+                    return Some(result);
```

Comment:
It's kind of weird to read the `if let Some()` twice but I'm not sure of a cleaner way.

Reply:
Yeah (assuming you mean the duplication between the two `LookupMode` cases here), it could be done with a closure, but this seemed fine for now...

Reply:
Another approach is to do

```rust
let candidates: &mut dyn _ = 
```

Oh, wait, even more horrible

```rust
for derefed_ty in autoderef::autoderef(db, resolver, ty.clone()) {
    if let Some(result) =
        iterate_inherent_methods(&derefed_ty, db, name, mode, krate, &mut callback)
    {
        return Some(result);
    }
    if let Some(result) =
        iterate_trait_method_candidates(&derefed_ty, db, resolver, name, mode, &mut callback)
    {
        return Some(result);
    }
    if let LookupMode::Path = mode {
        break;
    }
}
```

Reply:
I think I'll need to refactor this a bit anyway to get the method resolution order correct, so I'd leave it like this for now :smile: